### ISSUE TITLE
`nan` in gradients of a function, whose output is a itself a function of an optimal transport plan depending on the inputs: switching from `jnp.float32` to `jnp.float64` can help

### ISSUE THREAD

**Issue by theouscidda6** [ott-jax/ott#22]
I encountered a problem in optimizing a function `loss` whose output is itself a function of the optimal transport plan between two measurements defined from the inputs of `loss`.

Indeed, while the value of `loss` did not diverge and the norm of the gradients did not explode, I obtained from a certain moment `nan` in the gradients. 

All these computations were done with `DeviceArray` containing `jnp.float32`. I then switched to `jnp.float64` and there were no more nans, all calculations were done correctly.

I'm not sure why this change of type solved the problem (maybe you can help me with this?) but I think that switching from `jnp.float32` to `jnp.float64` can help some users!

**Remark**: `jax` by default enforces single-precision numbers to mitigate the Numpy API’s tendency to aggressively promote operands to double, you need to run th following code to allow the switching to `jnp.float64`:
```
from jax.config import config
config.update("jax_enable_x64", True)
```
Ref: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html

**Comment by marcocuturi** [comment]
Thanks Théo for these comments! are you running all your computations in float 64 since? did you encounter another problem later on?

**Comment by marcocuturi** [comment]
Thanks Théo!
